src/edu/washington/escience/myria/operator/agg/StreamingAggregate.java

```diff
@@ -0,0 +1,275 @@
+          /* Different grouping key than current one, flush current agg result to result buffer. */
+          addToResult();
+          // store current group key as a tuple
+          for (int gKey = 0; gKey < gFields.length; ++gKey) {
+            switch (gTypes[gKey]) {
+              case BOOLEAN_TYPE:
+                curGroupKey.set(gKey, tb.getBoolean(gFields[gKey], row));
+                break;
+              case STRING_TYPE:
+                curGroupKey.set(gKey, tb.getString(gFields[gKey], row));
+                break;
+              case DATETIME_TYPE:
+                curGroupKey.set(gKey, tb.getDateTime(gFields[gKey], row));
+                break;
+              case INT_TYPE:
+                curGroupKey.set(gKey, tb.getInt(gFields[gKey], row));
+                break;
+              case LONG_TYPE:
+                curGroupKey.set(gKey, tb.getLong(gFields[gKey], row));
+                break;
+              case FLOAT_TYPE:
+                curGroupKey.set(gKey, tb.getFloat(gFields[gKey], row));
+                break;
+              case DOUBLE_TYPE:
+                curGroupKey.set(gKey, tb.getDouble(gFields[gKey], row));
+                break;
+            }
+          }
+          reinitializeAggStates();
+        }
+        // update aggregator states with current tuple
+        for (int agg = 0; agg < aggregators.length; ++agg) {
+          aggregators[agg].addRow(tb, row, aggregatorStates[agg]);
+        }
+      }
+      tb = child.nextReady();
+    }
+
+    /*
+     * We know that child.nextReady() has returned <code>null</code>, so we have processed all tuple we can. Child is
+     * either EOS or we have to wait for more data.
+     */
+    if (child.eos()) {
+      addToResult();
+      return getResultBatch();
+    }
+    return null;
+  }
+
+  /**
+   * Re-initialize aggregator states for new group key.
+   *
+   * @throws DbException if any error
+   */
+  private void reinitializeAggStates() throws DbException {
+    aggregatorStates = null;
+    aggregatorStates = AggUtils.allocateAggStates(aggregators);
+  }
+
+  /**
+   * Add aggregate results with previous grouping key to result buffer.
+   *
+   * @throws DbException if any error
+   */
+  private void addToResult() throws DbException {
+    int fromIndex = 0;
+    for (; fromIndex < curGroupKey.numColumns(); ++fromIndex) {
+      TupleUtils.copyValue(curGroupKey, fromIndex, 0, resultBuffer, fromIndex);
+    }
+    for (int agg = 0; agg < aggregators.length; ++agg) {
+      aggregators[agg].getResult(resultBuffer, fromIndex, aggregatorStates[agg]);
+      fromIndex += aggregators[agg].getResultSchema().numColumns();
+    }
+  }
+
+  /**
+   * @return A batch's worth of result tuples from this aggregate.
+   * @throws DbException if there is an error.
+   */
+  private TupleBatch getResultBatch() throws DbException {
+    Preconditions.checkState(getChild().eos(), "cannot extract results from an aggregate until child has reached EOS");
```

Comment:
I don't understand this at all. I thought the whole point of a "streaming aggregator" was that you *could* return aggregation results before the child reached EOS. Am I missing something? (I assume that you're familiar with the difference in semantics between `TupleBuffer` and `TupleBatchBuffer`. I wonder if you're having trouble with the fact that you can't extract a `TupleBatch` from a `TupleBuffer` until it's finalized. If so, the answer is to use `TupleBatchBuffer` unless you need random access to the rows in the buffer.)